### PR TITLE
Update parameter update doc.

### PR DESCRIPTION
**Pull request by dan-zheng**
Update "Parameter Update" documentation using `KeyPathIterable` and
`Differentiable`.

Add "evolution from previous design" section.

### CODE REVIEW

**Comment by rxwei** [comment]
LGTM overall. How about a different name that mentions "optimizers" somewhere?

"Hierarchical parameter iteration and optimizations"?

**Comment by rxwei** [comment]
Don't forget to also add a link on README.

**Comment by dan-zheng** [comment]
> Don't forget to also add a link on README.

Done in https://github.com/tensorflow/swift/pull/136/commits/bd94d004790a53eb633adaa37bd875bdb0bb01c1.
Added some other updates in https://github.com/tensorflow/swift/pull/136/commits/b4907fc794e50b93f68020490225636548f3be80.
Ready for re-review!

File: README.md [hunk 1]
```diff
@@ -139,8 +139,8 @@ out the following documentation.
 - [Automatic Differentiation Whitepaper](docs/AutomaticDifferentiation.md)
 - [Automatic Differentiation Manifesto](https://gist.github.com/rxwei/30ba75ce092ab3b0dce4bde1fc2c9f1d)
 - [Dynamic Property Iteration using Key Paths](docs/DynamicPropertyIteration.md)
+- [Hierarchical parameter iteration and optimization](docs/ParameterOptimization.md)
```
Comment (rxwei) [inline README.md]: Make this match the capitalization of other links.

File: docs/ParameterOptimization.md [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+However, in Swift, this is difficult. Constructing a collection of parameters is difficult because parameters aren't required to have the same type: for example, a model may have parameters with types `Float`, `Tensor<Float>`, and `Tensor<Double>`. To represent parameters as a collection, advanced type-erasure is needed to generalize all parameter types. It's also not clear how parameter update would work (at compile-time, how is it possible to identify the correct concrete `+` function for a type-erased parameter value?).
+
+Additionally, while we want to enable code that achieves the following:
+```swift
+for (inout θ, dθ) in zip(parameters, gradients) {
+    θ -= learningRate * dθ
+}
+```
+
+We don't want to actually lower the for-loop or zip operation to TensorFlow (lowering wouldn't be straightforward or and lowered representation wouldn't be efficient). Instead, we want to fully unroll the loop into individual straight-line statements:
```
Comment (dan-zheng) [inline docs/ParameterOptimization.md]: > We don't want to actually lower the for-loop or zip operation to TensorFlow (lowering wouldn't be straightforward or and lowered representation wouldn't be efficient). Instead, we want to fully unroll the loop into individual straight-line statements:

Note:  this efficiency concern was addressed later in the old version of this doc, but not in this updated version.

**Pull request merged**